src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs

```diff
@@ -66,10 +66,16 @@ public async Task SynchronizeWithBuildAsync(ImmutableDictionary<ProjectId, Immut
                     RaiseProjectDiagnosticsIfNeeded(project, stateSets, oldAnalysisData.Result, newResult);
                 }
 
-                // if we have updated errors, refresh open files
-                if (buildDiagnostics.Count > 0 && PreferLiveErrorsOnOpenedFiles(options))
+                // Refresh diagnostics for open files after solution build completes.
+                if (onBuildCompleted && PreferLiveErrorsOnOpenedFiles(options))
                 {
-                    // enqueue re-analysis of open documents.
+                    // Enqueue re-analysis of active document.
+                    if (_documentTrackingService?.GetActiveDocument(solution) is { } activeDocument)
+                    {
+                        AnalyzerService.Reanalyze(Workspace, documentIds: ImmutableArray.Create(activeDocument.Id), highPriority: true);
+                    }
```

Comment:
if you ant to be very slick, you can even then do this in three phases:

1. active doc.
2. visible docs.
3. open docs.

Reply:
Yes, it is possible. I am not sure if we need this much of micro-optimization though - changing active document should refresh the diagnostics.